### PR TITLE
Register binding for newly created vars for commonjs transforms

### LINKED ISSUE #13665: [Bug]: `scope.hasBinding` and `scope.getBinding` inconsistent for newly created vars
### 💻

- [x] Would you like to work on a fix?

### How are you using Babel?

Programmatic API (`babel.transform`, `babel.parse`)

### Input code

```js
const babel = require('@babel/core');

const myPlugin = () => ({
  visitor: {
    Program: {
      exit(programPath) {
        // Sub-traversal to ensure runs after ESM->CJS transform
        programPath.traverse({
          ReferencedIdentifier(path) {
            if (!path.findParent(p => p.isFunction())) return;
            const varName = path.node.name;
            path.addComment(
              'leading',
              ` ${varName} ${
                path.scope.hasBinding(varName) ? 'hasBinding' : 'no hasBinding'
              }, ${
                path.scope.getBinding(varName) ? 'getBinding' : 'no getBinding'
              } `
            );
          }
        });
      }
    }
  }
});

const input = `
  import {x} from './foo.js';
  var y = true;
  function f() {
    return [x, y, console];
  }
`;

const output = babel.transformSync(input, {
  plugins: [
    '@babel/plugin-transform-modules-commonjs',
    myPlugin
  ]
}).code;

console.log(output);
```

### Configuration file name

n/a

### Configuration

n/a

### Current and expected behavior

In above example, [@babel/plugin-transform-modules-commonjs](https://babeljs.io/docs/en/babel-plugin-transform-modules-commonjs) transforms `import {x} from './foo.js';` to `var _foo = require("./foo.js");` and `x` to `_foo.x`.

My plugin runs after this transform and checks whether the vars used inside function `f` are bound to a variable definition in the file, and if so, to where.

All works as expected, except for the newly-created `_foo` variable. In that case, `path.scope.hasBinding()` returns `true` (correctly), but `path.scope.getBinding()` returns `undefined` (incorrectly).

The output from above example is:

```js
"use strict";
var _foo = require("./foo.js");
var y = true;
function f() {
  return [
    /* _foo hasBinding, no getBinding */ _foo.x,
    /* y hasBinding, getBinding */ y,
    /* console no hasBinding, no getBinding */ console
  ];
}
```

I would expect the comment before `_foo.x` to read: `_foo hasBinding, getBinding`.

### Environment

```
System:
  OS: macOS Mojave 10.14.6
Binaries:
  Node: 16.6.1 - ~/.nvm/versions/node/v16.6.1/bin/node
  npm: 7.20.3 - ~/.nvm/versions/node/v16.6.1/bin/npm
npmPackages:
  @babel/core: ^7.15.0 => 7.15.0 
  @babel/generator: ^7.15.0 => 7.15.0 
  @babel/helper-plugin-utils: ^7.14.5 => 7.14.5 
  @babel/parser: ^7.15.0 => 7.15.0 
  @babel/plugin-transform-arrow-functions: ^7.14.5 => 7.14.5 
  @babel/plugin-transform-modules-commonjs: ^7.15.0 => 7.15.0 
  @babel/plugin-transform-react-jsx: ^7.14.9 => 7.14.9 
  @babel/plugin-transform-strict-mode: ^7.14.5 => 7.14.5 
  @babel/register: ^7.14.5 => 7.14.5 
  @babel/traverse: ^7.15.0 => 7.15.0 
  @babel/types: ^7.15.0 => 7.15.0 
  babel-jest: ^27.0.6 => 27.0.6 
  babel-plugin-dynamic-import-node: ^2.3.3 => 2.3.3 
  eslint: ^7.32.0 => 7.32.0 
  jest: ^27.0.6 => 27.0.6
```

### Possible solution

I am actually not entirely sure if this is intended behavior or not, but it certainly seems counter-intuitive.

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by The-x-Theorist**
| Q                        | A 
| ------------------------ | ---
| Fixed Issues?            | Fixes #13665  !-- remove the (`) quotes and write "Fixes" before the number to link the issues -->
| Patch: Bug Fix?          |
| Major: Breaking Change?  |
| Minor: New Feature?      |
| Tests Added + Pass?      | Yes
| Documentation PR Link    | 
| Any Dependency Changes?  |
| License                  | MIT

Newly created vars will get registered as soon as `path.unshiftContainer()` happens.

<a href="https://gitpod.io/#https://github.com/babel/babel/pull/14097"></a>

### CODE REVIEW

**Comment by overlookmotel** [comment]
Nice one! Thanks loads for doing this.

I guess, strictly speaking, should also make sure that old bindings are also removed. e.g.:

```js
// Input
import React from "react";

// Output
var _react = require("react");
```

So, as well as the new `_react` binding being created, the old `React` binding should be destroyed.

Also, vars which are used in CommonJS are renamed:

```js
// Input
let module, exports, require, __dirname, __filename;

// Output
let _module, _exports, _require, _dirname, _filename;
```

Do bindings for these var name changes get dealt with too? (maybe they do automatically as they're just variable renaming not creation of new vars, I'm not sure)

I don't know how difficult it would be to handle the above. Certainly your PR as it is already is a big improvement on before.

**Comment by The-x-Theorist** [comment]
@overlookmotel Earlier there were no bindings registered for the newly created vars as of this, I think that there is no way we can remove the old bindings which weren't present actually.

**Comment by overlookmotel** [comment]
Ah sorry I wasn't clear. The `React` variable exists in the original source code, and so would have a binding (that's what I mean by the "old" binding). After the plugin has run, that var no longer exists, so should not have a binding any more.

I think the binding to non-existent variable `React` will still exist though, unless it's manually deleted.

**Comment by The-x-Theorist** [comment]
@overlookmotel Now I think I understand, the old bindings should be deleted which were present before the plugin ran, like for React variable, Right?

**Comment by The-x-Theorist** [comment]
@overlookmotel Can you suggest a way with help of which oldBindings can be deleted?

**Comment by overlookmotel** [comment]
Yes, that's exactly what I meant.

I'm afraid no, I can't offer any suggestions. I'm not actually very familiar with the code in Babel which deals with scopes/bindings. Maybe one of the maintainers can offer some guidance.

Or it may be that it's fine to leave it as it is. The biggest fault was that the new bindings weren't registered (which you've solved). It's of lesser importance that the old bindings are removed.

File: packages/babel-plugin-transform-modules-commonjs/src/index.ts [hunk 1]
```diff
@@ -235,6 +235,12 @@ export default declare((api, options) => {
 
           ensureStatementsHoisted(headers);
           path.unshiftContainer("body", headers);
+          path.get("body").forEach(path => {
+            if (headers.indexOf(path.node) === -1) return;
+            if (path.isVariableDeclaration()) {
+              path.scope.registerDeclaration(path);
+            }
```
Comment (nicolo-ribaudo) [inline packages/babel-plugin-transform-modules-commonjs/src/index.ts:242]: Note for reviewers: this code is the same as https://github.com/babel/babel/blob/bee3e358515d0f3a67c557fad9a9afa0fe929a5e/packages/babel-core/src/transformation/file/file.ts#L217-L223

**Review by nicolo-ribaudo — APPROVED** [review]
We should definitely remove the old bindings, but it can be done in a separate PR!

**Review by existentialism — APPROVED** [review]

**Review by JLHwung — APPROVED** [review]

**Comment by nicolo-ribaudo** [comment]
Oh, it looks like I already have a PR for the "remove old bindings" part: https://github.com/babel/babel/pull/10640

**Pull request merged**